core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1417,12 +1421,15 @@ object Parser extends ParserInstances {
           unmap(p1) match {
             case Prod(p11, p12) =>
               // right associate so
-              // we can find check matches a bit faster
-              Prod1(p11, unmap(Parser.product(p12, p2)))
+              // we can check matches a bit faster
+              // note: p12 is already unmapped, so
+              // we wrap with Void to prevent n^2 cost
+              Prod1(p11, unmap(Parser.product(p12.void, p2)))
```

Comment:
we use `.void` here because if `p12` is a parser1 but `p2` isn't we need to make sure to keep the `p12.void` to be a Parser1 and .void does this. Otherwise we would need to test here.